src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -15,325 +15,285 @@
+    } else {
+        return; // asserted above
     }
 
-    // when "concurrency width" is implemented this will be decremented instead
-    self.runningWorkItemCount = 0;
-    [self _callNextReadyWorkItem];
-    os_unfair_lock_unlock(&_lock);
+    // Always dispatch even if there is no readyHandler as this avoids synchronously
+    // re-entering the MTRAsyncWorkQueueCode, simplifying the implementation.
+    auto readyHandler = _readyHandler;
+    dispatch_async(_queue, ^{
+        if (readyHandler) {
+            readyHandler(context, retryCount, completion);
+        } else {
+            completion(MTRAsyncWorkComplete);
+        }
+    });
 }
 
-- (void)endWork:(MTRAsyncWorkItem *)workItem
+- (BOOL)isComplete
 {
-    [self _postProcessWorkItem:workItem retry:NO];
+    return _state == MTRAsyncWorkItemComplete;
 }
 
-- (void)retryWork:(MTRAsyncWorkItem *)workItem
+- (void)markComplete
 {
-    [self _postProcessWorkItem:workItem retry:YES];
+    NSAssert(_state >= MTRAsyncWorkItemEnqueued, @"work item was not enqueued (%ld)", (long) _state);
+    _state = MTRAsyncWorkItemComplete;
 }
 
-// assume lock is held while calling this
-- (void)_callNextReadyWorkItem
+- (void)cancel
 {
-    // when "concurrency width" is implemented this will be checked against the width
-    if (self.runningWorkItemCount) {
-        // can't run next work item until the current one is done
-        return;
-    }
-
-    // only proceed to mark queue as running if there are items to run
-    if (self.items.count) {
-        // when "concurrency width" is implemented this will be incremented instead
-        self.runningWorkItemCount = 1;
-
-        MTRAsyncWorkItem * workItem = self.items.firstObject;
-
-        // Check if batching is possible or needed. Only ask work item to batch once for simplicity
-        if (workItem.batchable && workItem.batchingHandler && (workItem.retryCount == 0)) {
-            while (self.items.count >= 2) {
-                MTRAsyncWorkItem * nextWorkItem = self.items[1];
-                if (!nextWorkItem.batchable || (nextWorkItem.batchingID != workItem.batchingID)) {
-                    // next item is not eligible to merge with this one
-                    break;
-                }
-
-                BOOL fullyMerged = NO;
-                workItem.batchingHandler(workItem.batchableData, nextWorkItem.batchableData, &fullyMerged);
-                if (!fullyMerged) {
-                    // We can't remove the next work item, so we can't merge anything else into this one.
-                    break;
-                }
-
-                [self.items removeObjectAtIndex:1];
-            }
+    if (_state != MTRAsyncWorkItemComplete) {
+        _state = MTRAsyncWorkItemComplete;
+        auto cancelHandler = _cancelHandler;
+        if (cancelHandler) {
+            // Note that this does not prevent a race against
+            // the readyHandler calling the work completion.
```

Comment:
So what (I think) I have implemented is that once a WorkItem has been enqueued (i.e. is owned by the queue) all state changes happen under the queue lock. So there is no race between `_callNextReadyWorkItem` and `invalidate` (and `invalidate` is the only way to cancel a work item currently). Cancellation can of course happen after the readyHandler has already been called (or dispatched to be called), but that's existing and expected behavior to cancel (if possible) a job that's already executing.

The race is due to the possible delay between the queue deciding to cancel a work item and the cancelHandler call actually being run on the dispatch queue. It's possible for the work completion handler to be called in-between these two points in time -- though it will return `NO` because from the perspective of the queue the work is already cancelled, so it's possible for the work code to deal with this.

This race also exists in the existing code; the decision to call the cancelHandler is made by code under the queue and work item locks, and then the dispatch is performed. However this is all academic, because no existing work we have even registers a cancel handler.

Reply:
> The race is due to the possible delay between the queue deciding to cancel a work item and the cancelHandler call actually being run on the dispatch queue. 

Right.

> This race also exists in the existing code

That's what I said above, yes.  It had a smaller window, since you have to race against linear code, not async-dispatched code, but it was still there.

So the thing is: having a work item do any of its "ready" work after invalidation is pretty dangerous.  It can lead to things being queued on the now-quiescent Matter work queue, where they lie in wait for the next startup.

Looking at how this can be reached (via `invalidate` only, as you note), it looks like it's called on an arbitrary thread/queue (wherever the SDK consumer decides to call `shutdown` on a controller).  

The real invariant we need to preserve is that no ready handler runs after invalidate returns.

It looks like `callReadyHandlerWithContext` and `_callNextReadyWorkItem` are both called with the lock held, right?  I guess that's pre-existing, though it does mean we are calling out into the consumer-provided batching handler with the lock held, which gives me the willies.  But if we want to make sure the blocks callReadyHandlerWithContext submits never run after invalidate, what are our options?

We could have that block close over a weak-self, stongify, take the lock and check whether we have been invalidated or something.  But that would not help the situation when we are in the middle of running a readyHandler already when the invalidate comes in.

Maybe we need a followup for this, because this is sounding hard.  But I suspect we will be getting random crashes for a while if we don't actually end up with something where stack shutdown effectively stops all ongoing work and joins the threads....

Reply:
If we can assume invalidate is called on the same queue as the work items, then the async could be constructed as a `dispatch_block_create` that is set on the work item. The invalidate call would call `dispatch_block_cancel` on it, and we'd be done.

If we cannot assume invalidate is called, then we may need the above plus something `invalidate` can wait on, if we need nothing to be running after invalidate returns.

Reply:
Sadly, the invalidate is called on the framework consumer's queue (when they shut down the controller), but the items in practice run on `"org.csa-iot.matter.framework.device.workqueue"`.

Reply:
Note that we can also shut down the device's work queue when the device is invalidated, but that does not help if an item is running on that queue right at that second....

On the other hand, maybe this is a non-issue because almost all of our dispatch to the Matter queue should be going via the controller, and that is getting shut down...  So once it gets far enough we can't dispatch on that queue anymore.  The exceptions are in `MTRBaseSubscriptionCallback`

Reply:
Actually, I just tested this and we could do `dispatch_block_cancel` and then `dispatch_block_wait` if the invalidate comes from a different queue, to wait until it finishes executing, if the block already started running.

Reply:
And I guess the operational browser is an exception too.  But I really hope that `DNSServiceRefDeallocate` will prevent the blocks from the DNS stuff from running after that, and we do that on the Matter queue, so it should be safe.

Reply:
I think I have convinced myself this is safe now.